src/projectDB/projectversioning.cpp

```diff
@@ -0,0 +1,257 @@
+{
+    int majorDiff = _major - other.major();
+    int minorDiff = _minor - other.minor();
+    int patchDiff = _patch - other.patch();
+    return majorDiff + minorDiff + patchDiff;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+pair<string, string> extractVersionInfoFromTag(const string tag, int* aheadBy)
+{
+    regex rgx("v((\\d+\\.\\d+\\.\\d+)(-(beta|alpha))?(\\.(\\d)*)?(-(\\d+))?)");
+    smatch matches;
+    pair<string, string> found;
+
+    if (regex_search(tag, matches, rgx)) {
+        // captured group for version should exist on index 2
+        found.first = matches[2].str();
+
+        // captured groups for alpha/beta build stage should be on index 4 and 5
+        found.second = matches[4].str() + matches[5].str();
+
+        // captured group for commit count should be on index 8
+        string commitCountStr = matches[8].str();
+        if (aheadBy != nullptr && commitCountStr.size() > 0) {
+            *aheadBy = stoi(commitCountStr);
+        } else if (aheadBy != nullptr) {
+            *aheadBy = 0;
+        }
+    } else {
+        cerr << "Error: given tag string \""
+             << tag
+             << "\" does not have expected format"
+             << endl;
+    }
+
+    return found;
+}
+
+int getDbVersionForApp(const Version& currentVersion)
+{
+    auto oldestVersion = begin(appDbVersionMap)->first;
+    if (currentVersion < oldestVersion)
+        return -1;
+
+    Version olderVersion;
+    Version laterVersion;
+    for (const auto entry : appDbVersionMap) {
+        auto version = entry.first;
+        if (version < currentVersion) {
+            olderVersion = version;
+        } else {
+            laterVersion = version;
+            break;
+        }
+    }
+    if (currentVersion == laterVersion)
+        return appDbVersionMap.at(laterVersion);
+
+    return appDbVersionMap.at(olderVersion);
+}
+
+int getLatestDbVersion()
+{
+    if (!appDbVersionMap.empty()) {
+        auto lastElement = --end(appDbVersionMap);
+        return lastElement->second;
+    }
+    return -1;
+}
+
+string generateUpgradeScript(const int fromVersion, const int toVersion)
+{
+    // the assumption here is that every database version must exist as a key
+    if (!dbVersionUpgradeScripts.count(fromVersion)
+        || !dbVersionUpgradeScripts.count(toVersion - 1)) {
+        cerr << "No conversion script can be generated for: "
+             << fromVersion << " → " << toVersion
+             << endl;
+        return "";
+    }
+
+    string mergedUpgradeScript;
+    for (const auto entry : dbVersionUpgradeScripts) {
+        int dbVersion = entry.first;
+        string upgradeScript = entry.second;
+
+        // relying on keys of map being sorted for generating the correct script
+        if (dbVersion >= fromVersion && dbVersion < toVersion)
+            mergedUpgradeScript += upgradeScript + "\n\n";
+    }
+
+    return mergedUpgradeScript;
+}
```

Comment:
Not sure what's a good way to implement this. 
0->2 can contain a list of upgrade scripts from 0-1 and 1-2, so we won't have to manually write them again in such cases

Reply:
Anyway, I am resolving this for now. Not an immediate requirement

Reply:
But then instead of having a list of such scripts, I am using the `generateUpgradeScript` method to create it when needed. That way we avoid having maintaining such lists.